diff -Naur freetype-2.5.0.1.orig/builds/unix/config.sub freetype-2.5.0.1/builds/unix/config.sub
--- freetype-2.5.0.1.orig/builds/unix/config.sub	2013-06-19 23:29:27.000000000 +0200
+++ freetype-2.5.0.1/builds/unix/config.sub	2013-06-26 14:39:49.516498064 +0200
@@ -634,6 +634,10 @@
 		basic_machine=elxsi-elxsi
 		os=-bsd
 		;;
+	emscripten)
+		basic_machine=le32-unknown
+		os=-emscripten
+		;;
 	encore | umax | mmax)
 		basic_machine=ns32k-encore
 		;;
@@ -1506,6 +1510,8 @@
 		;;
 	-nacl*)
 		;;
+	-emscripten*)
+		;;
 	-none)
 		;;
 	*)
