diff -Naur freetype-2.4.11.orig/builds/unix/config.sub freetype-2.4.11/builds/unix/config.sub
--- freetype-2.4.11.orig/builds/unix/config.sub	2012-12-20 08:35:22.000000000 +0100
+++ freetype-2.4.11/builds/unix/config.sub	2013-01-11 16:41:42.002166917 +0100
@@ -639,6 +639,10 @@
 		basic_machine=elxsi-elxsi
 		os=-bsd
 		;;
+	emscripten)
+		basic_machine=le32-unknown
+		os=-emscripten
+		;;
 	encore | umax | mmax)
 		basic_machine=ns32k-encore
 		;;
@@ -1510,6 +1514,8 @@
 		;;
 	-nacl*)
 		;;
+	-emscripten*)
+		;;
 	-none)
 		;;
 	*)
