diff -Naur libogg-1.3.1.orig/config.sub libogg-1.3.1/config.sub
--- libogg-1.3.1.orig/config.sub	2012-09-18 04:53:51.000000000 +0200
+++ libogg-1.3.1/config.sub	2013-06-16 17:45:50.374252271 +0200
@@ -490,6 +490,10 @@
 		basic_machine=m68k-unknown
 		os=-sysv4
 		;;
+	android)
+		basic_machine=i686-pc
+		os=-android
+		;;
 	apollo68)
 		basic_machine=m68k-apollo
 		os=-sysv
@@ -630,6 +634,10 @@
 		basic_machine=elxsi-elxsi
 		os=-bsd
 		;;
+	emscripten)
+		basic_machine=i686-pc
+		os=-emscripten
+		;;
 	encore | umax | mmax)
 		basic_machine=ns32k-encore
 		;;
@@ -833,6 +841,10 @@
 		basic_machine=le32-unknown
 		os=-nacl
 		;;
+	nacl)
+		basic_machine=i686-pc
+		os=-nacl
+		;;
 	ncr3000)
 		basic_machine=i486-ncr
 		os=-sysv4
@@ -1362,7 +1374,8 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
+	      | -nacl* | -android* | -emscripten*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -Naur libogg-1.3.1.orig/configure libogg-1.3.1/configure
--- libogg-1.3.1.orig/configure	2013-05-20 16:00:34.000000000 +0200
+++ libogg-1.3.1/configure	2013-06-16 17:45:50.377585560 +0200
@@ -11645,7 +11645,7 @@
 		;;
         *)
                 DEBUG="-g -Wall -fsigned-char"
-                CFLAGS="-O20 -fsigned-char"
+                CFLAGS="-fsigned-char"
                 PROFILE="-O20 -g -pg -fsigned-char"
 		;;
         esac
