diff -Naur libpng-1.6.3.orig/CMakeLists.txt libpng-1.6.3/CMakeLists.txt
--- libpng-1.6.3.orig/CMakeLists.txt	2013-07-18 02:03:00.000000000 +0200
+++ libpng-1.6.3/CMakeLists.txt	2013-09-14 14:38:17.753681428 +0200
@@ -24,7 +24,7 @@
 find_package(ZLIB REQUIRED)
 include_directories(${ZLIB_INCLUDE_DIR})
 
-if(NOT WIN32)
+if(NOT WIN32 AND NOT EMSCRIPTEN)
   find_library(M_LIBRARY
     NAMES m
     PATHS /usr/lib /usr/local/lib
