diff -Naur lua-5.3.0.orig/src/liolib.c lua-5.3.0/src/liolib.c
--- lua-5.3.0.orig/src/liolib.c	2015-01-02 13:50:28.000000000 +0100
+++ lua-5.3.0/src/liolib.c	2015-02-13 20:08:24.438416851 +0100
@@ -412,8 +412,12 @@
 
 /* access to locale "radix character" (decimal point) */
 #if !defined(l_getlocaledecpoint)
+#ifdef __ANDROID__
+#define l_getlocaledecpoint()     '.'
+#else
 #define l_getlocaledecpoint()     (localeconv()->decimal_point[0])
 #endif
+#endif
 
 
 /*
diff -Naur lua-5.3.0.orig/src/llex.c lua-5.3.0/src/llex.c
--- lua-5.3.0.orig/src/llex.c	2014-11-14 17:06:09.000000000 +0100
+++ lua-5.3.0/src/llex.c	2015-02-13 20:09:04.025281097 +0100
@@ -222,8 +222,12 @@
 
 
 #if !defined(l_getlocaledecpoint)
+#ifdef __ANDROID__
+#define l_getlocaledecpoint()   '.'
+#else
 #define l_getlocaledecpoint()	(localeconv()->decimal_point[0])
 #endif
+#endif
 
 
 #define buff2num(b,o)	(luaO_str2num(luaZ_buffer(b), o) != 0)
