diff -Naur lua-5.2.2.orig/src/llex.c lua-5.2.2/src/llex.c
--- lua-5.2.2.orig/src/llex.c	2013-03-16 22:10:18.000000000 +0100
+++ lua-5.2.2/src/llex.c	2013-04-03 18:42:00.327551527 +0200
@@ -211,7 +211,7 @@
 */
 static void trydecpoint (LexState *ls, SemInfo *seminfo) {
   char old = ls->decpoint;
-  ls->decpoint = getlocaledecpoint();
+  ls->decpoint = '.';
   buffreplace(ls, old, ls->decpoint);  /* try new decimal separator */
   if (!buff2d(ls->buff, &seminfo->r)) {
     /* format error with correct decimal point: no more options */
