diff -r eae894fafea0 include/SDL_config_android.h
--- a/include/SDL_config_android.h	Sun Jun 16 12:00:54 2013 +0200
+++ b/include/SDL_config_android.h	Sun Jun 16 18:44:25 2013 +0200
@@ -43,6 +43,8 @@
 #define HAVE_MATH_H 1
 #define HAVE_SIGNAL_H   1
 
+#define HAVE_GCC_ATOMICS   1
+
 /* C library functions */
 #define HAVE_MALLOC 1
 #define HAVE_CALLOC 1
@@ -127,7 +129,6 @@
 
 /* Enable OpenGL ES */
 #define SDL_VIDEO_OPENGL_ES 1
-#define SDL_VIDEO_RENDER_OGL_ES 1
 #define SDL_VIDEO_RENDER_OGL_ES2    1
 
 /* Enable system power support */
