diff -r 1b5280cd5885 include/SDL_config_android.h
--- a/include/SDL_config_android.h	Sun Dec 09 17:56:19 2012 -0800
+++ b/include/SDL_config_android.h	Mon Dec 10 20:54:06 2012 +0100
@@ -42,6 +42,7 @@
 #define HAVE_CTYPE_H	1
 #define HAVE_MATH_H	1
 #define HAVE_SIGNAL_H	1
+#define HAVE_GCC_ATOMICS 1
 
 /* C library functions */
 #define HAVE_MALLOC	1
@@ -127,7 +128,6 @@
 
 /* Enable OpenGL ES */
 #define SDL_VIDEO_OPENGL_ES	1
-#define SDL_VIDEO_RENDER_OGL_ES	1
 #define SDL_VIDEO_RENDER_OGL_ES2	1
 
 /* Enable system power support */
