diff -Naur SDL2-2.0.3.orig/include/SDL_stdinc.h SDL2-2.0.3/include/SDL_stdinc.h
--- SDL2-2.0.3.orig/include/SDL_stdinc.h	2014-03-16 03:31:42.000000000 +0100
+++ SDL2-2.0.3/include/SDL_stdinc.h	2015-02-15 00:14:00.156717838 +0100
@@ -86,6 +86,9 @@
 #if defined(HAVE_ICONV) && defined(HAVE_ICONV_H)
 # include <iconv.h>
 #endif
+#ifdef __ANDROID__
+# include <wchar.h>
+#endif
 
 /**
  *  The number of elements in an array.
