diff -Naur SDL_ttf-2.0.11.orig/config.sub SDL_ttf-2.0.11/config.sub
--- SDL_ttf-2.0.11.orig/config.sub	2012-01-15 05:44:08.000000000 +0100
+++ SDL_ttf-2.0.11/config.sub	2013-06-16 19:10:17.776996347 +0200
@@ -237,6 +237,18 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+	-android)
+		basic_machine=i686-pc
+		os=-android
+		;;
+	-nacl)
+		basic_machine=i686-pc
+		os=-nacl
+		;;
+	-emscripten)
+		basic_machine=i686-pc
+		os=-emscripten
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
@@ -1225,6 +1237,10 @@
 	pmac | pmac-mpw)
 		basic_machine=powerpc-apple
 		;;
+	emscripten)
+		basic_machine=i686-pc
+		os=-emscripten
+		;;
 	*-unknown)
 		# Make sure to match an already-canonicalized machine name.
 		;;
@@ -1297,7 +1313,8 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
+	      | -android* | -nacl* | -emscripten*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -Naur SDL_ttf-2.0.11.orig/Makefile.in SDL_ttf-2.0.11/Makefile.in
--- SDL_ttf-2.0.11.orig/Makefile.in	2012-01-15 05:44:08.000000000 +0100
+++ SDL_ttf-2.0.11/Makefile.in	2013-06-16 19:10:17.780329701 +0200
@@ -37,7 +37,6 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-noinst_PROGRAMS = showfont$(EXEEXT) glfont$(EXEEXT)
 subdir = .
 DIST_COMMON = README $(am__configure_deps) \
 	$(libSDL_ttfinclude_HEADERS) $(srcdir)/Makefile.am \
