diff -Naur SDL_ttf/config.sub SDL_ttf-nacl/config.sub
--- SDL_ttf/config.sub	2012-10-14 23:05:03.208773178 +0200
+++ SDL_ttf-nacl/config.sub	2012-10-14 23:07:49.747793628 +0200
@@ -237,6 +237,14 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+	-android)
+		basic_machine=i686-pc
+		os=-android
+		;;
+	-nacl)
+		basic_machine=i686-pc
+		os=-android
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
@@ -1297,7 +1305,8 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
+	      | -android* | -nacl*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -Naur SDL_ttf/Makefile.in SDL_ttf-nacl/Makefile.in
--- SDL_ttf/Makefile.in	2012-10-14 23:05:03.202106551 +0200
+++ SDL_ttf-nacl/Makefile.in	2012-10-14 23:08:02.784382725 +0200
@@ -37,7 +37,6 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-noinst_PROGRAMS = showfont$(EXEEXT) glfont$(EXEEXT)
 subdir = .
 DIST_COMMON = README $(am__configure_deps) \
 	$(libSDL2_ttfinclude_HEADERS) $(srcdir)/Makefile.am \
