diff -r e68c3eb54b66 Makefile.in
--- a/Makefile.in	Tue Jun 11 00:08:37 2013 -0700
+++ b/Makefile.in	Sun Jun 16 19:09:21 2013 +0200
@@ -55,7 +55,6 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-noinst_PROGRAMS = showfont$(EXEEXT) glfont$(EXEEXT)
 subdir = .
 DIST_COMMON = $(srcdir)/Makefile.in $(srcdir)/Makefile.am \
 	$(top_srcdir)/configure $(am__configure_deps) \
diff -r e68c3eb54b66 config.sub
--- a/config.sub	Tue Jun 11 00:08:37 2013 -0700
+++ b/config.sub	Sun Jun 16 19:09:21 2013 +0200
@@ -237,6 +237,18 @@
 		basic_machine=m68k-atari
 		os=-mint
 		;;
+	-android)
+		basic_machine=i686-pc
+		os=-android
+		;;
+	-nacl)
+		basic_machine=i686-pc
+		os=-nacl
+		;;
+	-emscripten)
+		basic_machine=i686-pc
+		os=-emscripten
+		;;
 esac
 
 # Decode aliases for certain CPU-COMPANY combinations.
@@ -1225,6 +1237,10 @@
 	pmac | pmac-mpw)
 		basic_machine=powerpc-apple
 		;;
+	emscripten)
+		basic_machine=i686-pc
+		os=-emscripten
+		;;
 	*-unknown)
 		# Make sure to match an already-canonicalized machine name.
 		;;
@@ -1297,7 +1313,8 @@
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* \
+	      | -android* | -nacl* | -emscripten*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
